### PR TITLE
[Merged by Bors] - feat(meta/interactive): Allow `case` to match multiple cases

### PR DESCRIPTION
**Pull request by eric-wieser**
Taking an example from the test, this extends the `case` syntax to allow square brackets listing multiple cases:

```lean
example (α β : Type*) (i j : α ⊕ β) : ℕ :=
begin
  with_cases {cases hi : i; cases hj : j},
  case [sum.inl sum.inl : i j, sum.inr sum.inr : i j] {
    exact 1,
    exact 1,
  },
  case [sum.inl sum.inr : i j, sum.inr sum.inl : i j] {
    exact 2,
    exact 2,
  },
end
```

With the cursor inside the `{`, this shows all the goals that the cases match.
If the intent is to run exactly the same proof against all matching cases, `all_goals` can be used within the `{`.

It also improves the error message for when `case` has bad syntax, by:
* Only showing the state for the current failing case
* Using `case` instead of `solve1` in the error message
* Not wrapping and indenting the line

---

[Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/239415-metaprogramming-.2F.20tactics/topic/Allowing.20focussing.20on.20multiple.20goals.20in.20.60case.60/near/220358809)

### CODE REVIEW

File: library/init/meta/interactive.lean [hunk 1]
```diff
@@ -709,24 +723,40 @@ hypotheses to `x, y, z`. The renaming functionality would be available only if
 the goal has a case tag.
 -/
 meta def case (ns : parse ident_*) (ids : parse $ (tk ":" *> ident_*)?) (tac : itactic) : tactic unit :=
```
Comment (eric-wieser) [inline library/init/meta/interactive.lean:725]: I guess I forgot to transfer this from my working copy. My best guess is that I meant
```suggestion
meta def case (ns : parse ident_*) (args : case_arg_parser) (tac : itactic) : tactic unit :=
```

File: library/init/meta/interactive.lean [hunk 2]
```diff
@@ -708,25 +722,41 @@ would select the first goal (or the first goal with a case tag), renaming
...
-   | (case_tag.hyps _ new_hyp_names) := do
-       let num_new_hyps := new_hyp_names.length,
-       when (num_ids > num_new_hyps) $ fail format!
-         "Invalid `case`: You gave {num_ids} names, but the case introduces
-         {num_new_hyps} new hypotheses.",
-       let renamings := rb_map.of_list (new_hyp_names.zip ids),
-       propagate_tags $ tactic.rename_many renamings tt tt
-   end,
-   solve1 tac
+meta def case (ns : parse ident_*) (args : case_arg_parser) (tac : itactic) : tactic unit :=
```
Comment (eric-wieser) [inline library/init/meta/interactive.lean:725]:
```suggestion
meta def case (args : case_arg_parser) (tac : itactic) : tactic unit :=
```

File: library/init/meta/interactive.lean [hunk 3]
```diff
@@ -708,25 +722,41 @@ would select the first goal (or the first goal with a case tag), renaming
...
-   | (case_tag.hyps _ new_hyp_names) := do
-       let num_new_hyps := new_hyp_names.length,
-       when (num_ids > num_new_hyps) $ fail format!
-         "Invalid `case`: You gave {num_ids} names, but the case introduces
-         {num_new_hyps} new hypotheses.",
-       let renamings := rb_map.of_list (new_hyp_names.zip ids),
-       propagate_tags $ tactic.rename_many renamings tt tt
-   end,
-   solve1 tac
+meta def case (args : case_arg_parser) (tac : itactic) : tactic unit :=
```
Comment (eric-wieser) [inline library/init/meta/interactive.lean:725]:
```suggestion
meta def case (args : parse case_arg_parser) (tac : itactic) : tactic unit :=
```

File: library/init/meta/interactive.lean [hunk 4]
```diff
@@ -708,25 +722,41 @@ would select the first goal (or the first goal with a case tag), renaming
...
-   | (case_tag.hyps _ new_hyp_names) := do
-       let num_new_hyps := new_hyp_names.length,
-       when (num_ids > num_new_hyps) $ fail format!
-         "Invalid `case`: You gave {num_ids} names, but the case introduces
-         {num_new_hyps} new hypotheses.",
-       let renamings := rb_map.of_list (new_hyp_names.zip ids),
-       propagate_tags $ tactic.rename_many renamings tt tt
-   end,
-   solve1 tac
+meta def case (args : parse case_arg_parser) (tac : itactic) : tactic unit :=
```
Comment (eric-wieser) [inline library/init/meta/interactive.lean:725]:
```suggestion
meta def case (args : parse case_parser) (tac : itactic) : tactic unit :=
```

File: library/init/meta/interactive.lean [hunk 5]
```diff
@@ -708,25 +722,41 @@ would select the first goal (or the first goal with a case tag), renaming
...
+    let other_goals := goals.filter (≠ goal),
+    set_goals [goal],
+    match tag with
+    | (case_tag.pi _ num_args) := do
+      intro_lst ids,
+      when (num_ids < num_args) $ intron (num_args - num_ids)
+    | (case_tag.hyps _ new_hyp_names) := do
+        let num_new_hyps := new_hyp_names.length,
+        when (num_ids > num_new_hyps) $ fail format!
+          "Invalid `case'`: You gave {num_ids} names, but the case introduces
```
Comment (eric-wieser) [inline library/init/meta/interactive.lean:741]:
```suggestion
          "Invalid `case`: You gave {num_ids} names, but the case introduces
```

File: library/init/meta/interactive.lean [hunk 6]
```diff
@@ -708,25 +722,41 @@ would select the first goal (or the first goal with a case tag), renaming
...
+    set_goals [goal],
+    match tag with
+    | (case_tag.pi _ num_args) := do
+      intro_lst ids,
+      when (num_ids < num_args) $ intron (num_args - num_ids)
+    | (case_tag.hyps _ new_hyp_names) := do
+        let num_new_hyps := new_hyp_names.length,
+        when (num_ids > num_new_hyps) $ fail format!
+          "Invalid `case`: You gave {num_ids} names, but the case introduces
+          {num_new_hyps} new hypotheses.",
```
Comment (eric-wieser) [inline library/init/meta/interactive.lean:742]:
```suggestion
          "Invalid `case`: You gave {num_ids} names, but the case introduces " ++
          "{num_new_hyps} new hypotheses.",
```

File: tests/lean/case.lean.expected.out [hunk 7]
```diff
@@ -74,41 +74,21 @@ ys_tl : list ℕ
...
-2 goals
 case list.cons
 xs_hd : ℕ,
 xs_tl : list ℕ
 ⊢ ℕ
-
-case list.nil
-⊢ ℕ
 case.lean:62:2: error: Invalid `case`: You gave 4 names, but the case introduces
          3 new hypotheses.
```
Comment (eric-wieser) [inline tests/lean/case.lean.expected.out:82]:
```suggestion
case.lean:62:2: error: Invalid `case`: You gave 4 names, but the case introduces 3 new hypotheses.
```

File: tests/lean/case.lean.expected.out [hunk 8]
```diff
@@ -74,41 +74,21 @@ ys_tl : list ℕ
 case.lean:56:2: error: Invalid `case`: You gave 3 names, but the case introduces
          2 new hypotheses.
```
Comment (eric-wieser) [inline tests/lean/case.lean.expected.out:75]:
```suggestion
case.lean:56:2: error: Invalid `case`: You gave 3 names, but the case introduces 2 new hypotheses.
```

File: library/init/meta/interactive.lean [hunk 9]
```diff
@@ -708,25 +722,41 @@ would select the first goal (or the first goal with a case tag), renaming
...
+    set_goals [goal],
+    match tag with
+    | (case_tag.pi _ num_args) := do
+      intro_lst ids,
+      when (num_ids < num_args) $ intron (num_args - num_ids)
+    | (case_tag.hyps _ new_hyp_names) := do
+        let num_new_hyps := new_hyp_names.length,
+        when (num_ids > num_new_hyps) $ fail format!
+          "Invalid `case`: You gave {num_ids} names, but the case introduces " ++
+          "{num_new_hyps} new hypotheses.",
```
Comment (eric-wieser) [inline library/init/meta/interactive.lean:742]:
```suggestion
        when (num_ids > num_new_hyps) $ fail format! $
          "Invalid `case`: You gave {num_ids} names, but the case introduces " ++
          "{num_new_hyps} new hypotheses.",
```

**Comment by eric-wieser** [comment]
Looks like I did a really bad job of copying my working version over into lean core. Should all be fixed now.

**Comment by sorawee** [comment]
I don't know if this unreasonable or not, but it would be nice if this the function can be split into `case_core` and `case`, where `case_core` returns `tactic (list expr × list expr)` which is a pair of `unsolved_goals` and `remaining_goals`. The implementation of `focus_case` (as discussed in https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Multiple.20case) would then simply be:

```
do (unsolved_goals, remaining_goals) ← case_core args tac,
   tactic.set_goals (unsolved_goals ++ remaining_goals)
```

**Pull request closed without merging**